### PR TITLE
docs(aio): rename service worker files, update examples, move SW under Techniques

### PR DESCRIPTION
**Pull request by sreevani-ship-it**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Doesn't test service worker functionality.
Service worker is under Fundamentals.
Issue Number: N/A

## What is the new behavior?
Service worker is under Techniques.
Added e2e testing for service worker functionality.
Renamed service worker doc filenames.
Added server-side redirects.
Refactored examples.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
This is a follow-up to #20736.

### CODE REVIEW

**Comment by chembu** [comment]
@alxhub, @gkalpak: Please take a look at the updated SW contents.

**Comment by gkalpak** [comment]
The Travis failures seem valid. Can you please take a look?

File: aio/content/guide/service-worker-config.md [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+<hr />
 
 The `src/ngsw-config.json` configuration file specifies which files and data URLs the Angular 
 service worker should cache and how it should update the cached files and data. The 
 CLI processes the configuration file during `ng build --prod`. Manually, you can process 
 it with the `ngsw-config` tool:
 
 ```sh
-ngsw-config dist src/ngswn-config.json /base/href
+ngsw-config dist src/ngsw-config.json /base/href
```
Comment (alxhub) [inline aio/content/guide/service-worker-config.md]: Nice catch!

**Review by alxhub — APPROVED** [review]

**Review by gkalpak — CHANGES_REQUESTED** [review]
Most of the issues have been fixed. Yay 💯 
A couple of comments:

1. AFAICT, the app's services are still not included in tests (see https://github.com/angular/angular/pull/20736#discussion_r155291213).

2. The plunker example is currently broken for two reasons:

    1. `environments/environment.ts` is not included.
    2. SystemJS is not configured to retrieve `@angular/service-worker`.

    Ignore (i) for now. We probably have to fix the `PlunkerBuilder`. (_EDIT: Created #20992 to track it._)

    For (ii), you need to add a mapping for `@angular/service-worker` in [systemjs.config.js][1], [systemjs.config.web.js][2], and [systemjs.config.web.build.js][3]:

    ```js
    '@angular/service-worker': 'npm:@angular/service-worker/bundles/service-worker.umd.js',
    ```

[1]: https://github.com/ExpertSupport/angular/blob/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/tools/examples/shared/boilerplate/systemjs/src/systemjs.config.js#L17-L31
[2]: https://github.com/ExpertSupport/angular/blob/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/tools/examples/shared/boilerplate/systemjs/src/systemjs.config.web.js#L38-L52
[3]: https://github.com/ExpertSupport/angular/blob/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/tools/examples/shared/boilerplate/systemjs/src/systemjs.config.web.build.js#L41-L55

File: aio/content/guide/service-worker-communications.md [hunk 2]
```diff
@@ -1,7 +1,15 @@
-# Communicating with service workers
+# Service Worker Communication
 
 Importing `ServiceWorkerModule` into your `AppModule` doesn't just register the service worker, it also provides a few services you can use to interact with the service worker and control the caching of your app.
 
+#### Prerequisites
+
+A basic understanding of the following:
+* [Getting Started with Service Workers](guide/service-worker-intro).
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/guide/service-worker-communications.md]: Updated URL

File: aio/content/examples/service-worker-getting-started/src/app/app.component.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+})
+export class AppComponent {
+  title = 'Service Workers';
+  updateCheckText = '';
+
+  constructor(private update: SwUpdate) {}
+
+  updateCheck(): void {
+      this
+        .update
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/src/app/app.component.ts]: Updated

File: aio/content/examples/service-worker-getting-started/src/app/app.component.html [hunk 4]
```diff
@@ -5,6 +5,10 @@ <h1>
   </h1>
   <img width="300" alt="Angular Logo" src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNTAgMjUwIj4KICAgIDxwYXRoIGZpbGw9IiNERDAwMzEiIGQ9Ik0xMjUgMzBMMzEuOSA2My4ybDE0LjIgMTIzLjFMMTI1IDIzMGw3OC45LTQzLjcgMTQuMi0xMjMuMXoiIC8+CiAgICA8cGF0aCBmaWxsPSIjQzMwMDJGIiBkPSJNMTI1IDMwdjIyLjItLjFWMjMwbDc4LjktNDMuNyAxNC4yLTEyMy4xTDEyNSAzMHoiIC8+CiAgICA8cGF0aCAgZmlsbD0iI0ZGRkZGRiIgZD0iTTEyNSA1Mi4xTDY2LjggMTgyLjZoMjEuN2wxMS43LTI5LjJoNDkuNGwxMS43IDI5LjJIMTgzTDEyNSA1Mi4xem0xNyA4My4zaC0zNGwxNy00MC45IDE3IDQwLjl6IiAvPgogIDwvc3ZnPg==">
 </div>
+
+<button id="check" (click)="updateCheck()">Check for Update</button>
+<p id="checkResult" *ngIf="updateCheckText !== ''">{{updateCheckText}}</p>
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/src/app/app.component.html]: @kapunahelewong : please take a look

File: aio/content/examples/service-worker-getting-started/e2e/app.e2e-spec.ts [hunk 5]
```diff
@@ -34,4 +35,10 @@ describe('sw-example App', () => {
       });
   });
 
+   it('SwUpdate.checkForUpdate() should return a rejected promise', () => {
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/e2e/app.e2e-spec.ts]: Added comment

File: aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts [hunk 6]
```diff
@@ -1,8 +1,6 @@
 import { Injectable } from '@angular/core';
 import { SwUpdate } from '@angular/service-worker';
-
-import { Observable } from 'rxjs/Observable';
-import 'rxjs/add/observable/interval';
+import { interval } from 'rxjs/observable/interval';
 
 function promptUser(event): boolean {
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts]: Removed promptUser function

File: aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts [hunk 7]
```diff
@@ -1,8 +1,6 @@
 import { Injectable } from '@angular/core';
 import { SwUpdate } from '@angular/service-worker';
-
-import { Observable } from 'rxjs/Observable';
-import 'rxjs/add/observable/interval';
+import { interval } from 'rxjs/observable/interval';
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts]: Removing prompt user function makes this line just above the "interval" usage.

File: aio/content/examples/service-worker-getting-started/plnkr.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+    "description": "Service Worker",
+    "basePath": "src/",
+    "files":[
+      "!**/*.d.ts",
+      "!**/*.js",
+      "!**/*.[1,2].*"
```
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/plnkr.json]: Removed.

File: aio/content/examples/service-worker-getting-started/src/app/app.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/examples/service-worker-getting-started/src/app/app.component.html]: Hey @alxhub I think this `ngIf` is necessary so we can see the failure for the e2e test, right?

File: aio/content/examples/service-worker-getting-started/e2e/app.e2e-spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/examples/service-worker-getting-started/e2e/app.e2e-spec.ts]: I don't see it 😕

File: aio/content/examples/service-worker-getting-started/plnkr.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/examples/service-worker-getting-started/plnkr.json]: I still see it 😕

File: aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts]: It still has to be part of the `docregion` (so it shows up in the code snippet in the guide.

File: aio/content/examples/service-worker-getting-started/src/app/app.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/examples/service-worker-getting-started/src/app/app.component.html]: It shouldn't be 😕

**Comment by gkalpak** [comment]
I also noticed the example zip is broken, because `@angular/service-worker` is not included in dependencies. You need to add it to [package.json][1].

[1]: https://github.com/angular/angular/blob/6e2a8a2ba44b95636d975a249d52ebadc80d8347/aio/tools/example-zipper/customizer/package-json/base.json#L6-L17

**Comment by gkalpak** [comment]
A couple mor comments:

- [Here][1] we mention `http-server`, but it is not sufficiently explained. (What is it? Where does it come from? How to install it?)

- The example zip is broken (for several reasons):
  - `@angular/service-worker` is not included in dependencies. You need to add it to [package.json][2].
  - The version of `@angular/cli` included in devDependencies is not compatible with `@angular/service-worker@5`.
  - We do not have `serviceWorker` enabled in `.angular-cli.json` (so it will not copy the config to `dist/`).

(The last two bullet-points are related to the `ExampleZipper` utility, so don't worry about them for now. We need to fix them.)

[1]: https://github.com/ExpertSupport/angular/blame/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/content/guide/service-worker-getting-started.md#L101-L108
[2]: https://github.com/angular/angular/blob/6e2a8a2ba44b95636d975a249d52ebadc80d8347/aio/tools/example-zipper/customizer/package-json/base.json#L6-L17

File: aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts]: Ok.

File: aio/content/examples/service-worker-getting-started/e2e/app.e2e-spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/e2e/app.e2e-spec.ts]: Didn't push the changes yet, was replying to keep track of the fix.

File: aio/content/examples/service-worker-getting-started/src/app/app.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (chembu) [reply aio/content/examples/service-worker-getting-started/src/app/app.component.html]: Removed ngIf.

**Comment by jenniferfell** [comment]
Re comment by @gkalpak above re HTTP server. I agree we should be more clear about what's happening here. The reader can use any HTTP they want locally. Something like this:

Because ng serve does not work with service workers, you must use a separate HTTP server to test your project locally. You can use any HTTP server. The example below uses the http-server package from npm. To reduce the possibility of conflicts, test on a dedicated port.

Maybe link from "http-server package" to this: https://www.npmjs.com/package/http-server

**Comment by gkalpak** [comment]
I am marking this PR as blocked on #20165 (which converts live-examples from plnkr to stackblitz). #20165 needs to land first, then this one needs to be adapted to also use stackblitz (the process is straight forward - don't worry about it for now).

You can continue working on this PR as usual (just don't get it merged before #20165).

**Comment by gkalpak** [comment]
Except for the issues listed below, the rest LGTM 👍
The pending issues (not necessarily to be fixed in this PR):

1. The app's services are still not included in tests (see https://github.com/angular/angular/pull/20736#discussion_r155291213).

2. The plunker example is currently broken for several reasons, including:
    - `environments/environment.ts` is not included. (Fixed in #21013.)
    - SystemJS is not configured to retrieve `@angular/service-worker`.
        Requires to add a mapping for `@angular/service-worker` in [systemjs.config.js][1], [systemjs.config.web.js][2], and [systemjs.config.web.build.js][3].
    - `ngsw-worker.js` is not available in plnkr.

3. The example zip is currently broken for several reasons, including:
    - `@angular/service-worker` is not included in dependencies. Needs to be added it to [package.json][4].
    - The version of `@angular/cli` included in devDependencies is not compatible with `@angular/service-worker@5`.
    - We do not have `serviceWorker` enabled in `.angular-cli.json` (so it will not copy the config to `dist/`).

    (The last two bullet-points are related to the `ExampleZipper` utility.)

[1]: https://github.com/ExpertSupport/angular/blob/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/tools/examples/shared/boilerplate/systemjs/src/systemjs.config.js#L17-L31
[2]: https://github.com/ExpertSupport/angular/blob/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/tools/examples/shared/boilerplate/systemjs/src/systemjs.config.web.js#L38-L52
[3]: https://github.com/ExpertSupport/angular/blob/faa88ea3f7fba2afde5f82788d409b50f68d6a41/aio/tools/examples/shared/boilerplate/systemjs/src/systemjs.config.web.build.js#L41-L55
[4]: https://github.com/angular/angular/blob/6e2a8a2ba44b95636d975a249d52ebadc80d8347/aio/tools/example-zipper/customizer/package-json/base.json#L6-L17

**Review by gkalpak — APPROVED** [review]
As discussed "offline", a live-example probably doesn't make sense (since you need to update the hashes n `ngsw.json` on every file change) and the downloadable zip is currently broken (and requires aio infrastructure changes to work).

LGTM with the `<live-example>` removed from the Getting Started guide.
<sub>(The remaining issues can be addressed in separate PRs.)</sub>

**Comment by alxhub** [comment]
@jenniferfell waiting on your final approval.

**Comment by jenniferfell** [comment]
"The bad news is that it appears that one or more commits were authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request." 

This has been resolved.

**Review by jenniferfell — APPROVED** [review]

**Pull request merged**